File: src/main/java/org/gautelis/durga/plugins/RegexExtract.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Extracts named capture groups from a field using a regular expression and
 * stores the results in the payload.
 *
 * <p>Config syntax:
 * <pre>
 * "source=log_line;pattern=(?&lt;ip&gt;\\d+\\.\\d+\\.\\d+\\.\\d+)\\s+(?&lt;method&gt;\\w+);target=parsed"
 * </pre>
 *
 * <p>Parameters:
 * <ul>
 *   <li>{@code source} — dot-notation path to the source field (required)</li>
 *   <li>{@code pattern} — Java regex with named groups (required)</li>
 *   <li>{@code target} — dot-notation path to store extracted groups (optional, default: top-level)</li>
 *   <li>{@code all} — if "true", extract all matches (optional, default matches first only)</li>
 * </ul>
 */
public final class RegexExtract implements Plugin {

    @Override
    public byte[] execute(byte[] payload, String config) throws Exception {
        String payloadStr = Plugin.toString(payload);
        String source = null;
        String pattern = null;
        String target = null;
        boolean all = false;
        if (config != null && !config.isBlank()) {
            String[] parts = config.split(";");
            for (String part : parts) {
                part = part.trim();
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String key = part.substring(0, eq).trim();
                    String val = part.substring(eq + 1).trim();
                    switch (key) {
                        case "source" -> source = val;
                        case "pattern" -> pattern = val;
                        case "target" -> target = val;
                        case "all" -> all = "true".equalsIgnoreCase(val);
                    }
                }
            }
        }
        return Plugin.toBytes(extract(payloadStr, source, pattern, target, all));
    }

    private RegexExtract() {
    }

    public static String extract(String json, String sourceField, String regex, String targetPath, boolean findAll) {
        if (sourceField == null || regex == null) {
            return json;
        }
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }
        if (!input.isObject()) {
            return json;
        }

        JsonNode sourceNode = PipelinePlugin.fieldAt(input, sourceField);
        if (sourceNode == null || !sourceNode.isTextual()) {
            return json;
        }

        String text = sourceNode.asText();
        Pattern compiled = Pattern.compile(regex, findAll ? 0 : 0);

        ObjectNode output = input.deepCopy();
        Matcher matcher = compiled.matcher(text);

        if (findAll) {
            boolean found = false;
            while (matcher.find()) {
                found = true;
                applyGroups(mapper, output, matcher, targetPath);
            }
            if (!found) {
                return json;
            }
        } else {
            if (!matcher.find()) {
                return json;
            }
            applyGroups(mapper, output, matcher, targetPath);
        }

        return output.toString();
    }

    private static void applyGroups(ObjectMapper mapper, ObjectNode root, Matcher matcher, String targetPath) {
        ObjectNode target;
        if (targetPath != null && !targetPath.isBlank()) {
            JsonNode existing = PipelinePlugin.fieldAt(root, targetPath);
            if (existing != null && existing.isObject()) {
                target = (ObjectNode) existing;
            } else {
                target = mapper.createObjectNode();
                PipelinePlugin.setFieldAt(root, targetPath, target);
            }
        } else {
            target = root;
        }

        Matcher groupMatcher = Pattern.compile("\\(\\?<(\\w+)>").matcher(matcher.pattern().pattern());
        java.util.Set<String> groupNames = new java.util.LinkedHashSet<>();
        while (groupMatcher.find()) {
            groupNames.add(groupMatcher.group(1));
        }

        for (String name : groupNames) {
            try {
                String value = matcher.group(name);
                if (value != null) {
                    target.put(name, value);
                }
            } catch (IllegalArgumentException ignored) {
            }
        }

        if (groupNames.isEmpty()) {
            int groupCount = matcher.groupCount();
            for (int i = 1; i <= groupCount; i++) {
                String value = matcher.group(i);
                if (value != null) {
                    target.put("group" + i, value);
                }
            }
        }
    }
}

File: durga-rust/src/plugins/regex_extract.rs
//! Port of `org.gautelis.durga.plugins.RegexExtract` (plugin id
//! `regex-extract`).
//!
//! Extracts named capture groups from a source field into the payload. Config:
//! `source=log_line;pattern=(?<ip>\d+\.\d+\.\d+\.\d+)\s+(?<method>\w+);target=parsed;all=false`.
//! With no named groups, positional groups are stored as `group1`, `group2`, ...

use regex::Regex;
use serde_json::{Map, Value};

use crate::pipeline::{field_at, set_field_at};
use crate::plugin::{Plugin, PluginError};

#[derive(Default)]
pub struct RegexExtract;

impl RegexExtract {
    pub fn new() -> Self {
        RegexExtract
    }

    pub fn extract(
        json: &str,
        source_field: Option<&str>,
        regex: Option<&str>,
        target_path: Option<&str>,
        find_all: bool,
    ) -> Result<String, PluginError> {
        let (Some(source_field), Some(regex)) = (source_field, regex) else {
            return Ok(json.to_string());
        };
        let mut input: Value = serde_json::from_str(json)
            .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;
        if !input.is_object() {
            return Ok(json.to_string());
        }
        let text = match field_at(&input, source_field) {
            Some(Value::String(s)) => s.clone(),
            _ => return Ok(json.to_string()),
        };
        let re = Regex::new(regex)
            .map_err(|e| -> PluginError { format!("Invalid regex pattern: {e}").into() })?;
        let group_names: Vec<String> =
            re.capture_names().flatten().map(|s| s.to_string()).collect();

        let mut matched = false;
        if find_all {
            let all: Vec<_> = re.captures_iter(&text).collect();
            for caps in &all {
                matched = true;
                apply_groups(&mut input, caps, &group_names, target_path);
            }
        } else if let Some(caps) = re.captures(&text) {
            matched = true;
            apply_groups(&mut input, &caps, &group_names, target_path);
        }
        if !matched {
            return Ok(json.to_string());
        }
        Ok(input.to_string())
    }
}

fn apply_groups(
    root: &mut Value,
    caps: &regex::Captures,
    group_names: &[String],
    target_path: Option<&str>,
) {
    // Collect the entries to write, then place them under the target path.
    let mut entries: Vec<(String, String)> = Vec::new();
    if group_names.is_empty() {
        for i in 1..caps.len() {
            if let Some(m) = caps.get(i) {
                entries.push((format!("group{i}"), m.as_str().to_string()));
            }
        }
    } else {
        for name in group_names {
            if let Some(m) = caps.name(name) {
                entries.push((name.clone(), m.as_str().to_string()));
            }
        }
    }

    match target_path.filter(|p| !p.is_empty()) {
        Some(path) => {
            let mut obj = match field_at(root, path) {
                Some(Value::Object(existing)) => existing.clone(),
                _ => Map::new(),
            };
            for (k, v) in entries {
                obj.insert(k, Value::String(v));
            }
            set_field_at(root, path, Value::Object(obj));
        }
        None => {
            if let Value::Object(map) = root {
                for (k, v) in entries {
                    map.insert(k, Value::String(v));
                }
            }
        }
    }
}

impl Plugin for RegexExtract {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        let mut source = None;
        let mut pattern = None;
        let mut target = None;
        let mut all = false;
        for part in config.split(';') {
            let part = part.trim();
            let Some(eq) = part.find('=') else { continue };
            if eq == 0 {
                continue;
            }
            let key = part[..eq].trim();
            let val = part[eq + 1..].trim();
            match key {
                "source" => source = Some(val.to_string()),
                "pattern" => pattern = Some(val.to_string()),
                "target" => target = Some(val.to_string()),
                "all" => all = val.eq_ignore_ascii_case("true"),
                _ => {}
            }
        }
        Ok(Some(Self::extract(
            payload,
            source.as_deref(),
            pattern.as_deref(),
            target.as_deref(),
            all,
        )?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_named_groups_to_target() {
        let out = RegexExtract::new()
            .execute_str(
                r#"{"line":"192.168.0.1 GET"}"#,
                r"source=line;pattern=(?<ip>\d+\.\d+\.\d+\.\d+)\s+(?<method>\w+);target=parsed",
            )
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["parsed"]["ip"], "192.168.0.1");
        assert_eq!(v["parsed"]["method"], "GET");
    }

    #[test]
    fn positional_groups_when_unnamed() {
        let out = RegexExtract::new()
            .execute_str(r#"{"s":"ab12"}"#, r"source=s;pattern=([a-z]+)(\d+)")
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["group1"], "ab");
        assert_eq!(v["group2"], "12");
    }
}
